Add metadata and citation instructions to prepare first release

nilsvu opened the pull request:
## Proposed changes

This PR adds a `Metadata.yaml` file with machine-readable information about the code, including an author list. This information will be updated at every release (approx. monthly) and transmitted to Zenodo. We have discussed most of the information in this PR already in meetings and in the request-for-comment document that was circulated, but to give everyone who is listed here the opportunity to review their information I'll leave this PR open and in "draft" mode until **Dec 4th, 9am CET** (I realize it's a holiday week in the US right now, so if anyone needs more time to review their information please just let me know here or on Slack).

Except for minor rewordings here's what changed from the last circulation of the request-for-comment document that most people probably read:

- Instead of naming the author tiers _Major_, _Regular_ and _Minor_ I named them _Core_, _Developers_ and _Contributors_ because I would like to avoid diminishing anyone's contributions by classifying them as "minor" or "regular". The name "Developers" has some overlap but is not identical to the term "Developers" in the proposed SXS publication policies. I think that's OK and I clarify the relation to the publication policies like this in the description:

  > Authorship on the SpECTRE DOI does not directly entitle a person for authorship rights on other publications. Authorship rights on publications related to SpECTRE are governed by our publication policy.

  We can add a link to the publication policy if and when we add it to the repository.
- I added to the "Core" tier description:

  > For the time a Developer serves as Core developer they take on some project-management responsibilities, such as merging pull-requests.
- I removed the section on publishing only a subset of releases to Zenodo, as discussed last Monday. The current plan is to automatically publish releases approx. monthly and mirror them to Zenodo, so they have a DOI and can be referred to in papers. People should cite the latest version that they use in their paper. It's also OK to just cite the DOI they find when visiting spectre-code.org at the time they happen to look up this information.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add metadata file

Co-authored-by: Nils Deppe <nd357@cornell.edu>

nilsvu committed:
Read metadata in CMake

nilsvu committed:
Add citation instructions to docs and README

nilsvu committed:
Add docs on versioning and releases

nilsvu commented:
@10220 @carmaza @denyzamelchor @dongjun826 @ffoucart @fmahebert @geoffrey4444 @gsb76 @hrueter @isaaclegred @isha1810 @jennifersanchez @kidder @macedo22 @markscheel @MarloMo @mgiesler @moxcodes @nilsdeppe @osheamonn @prayush @sierracthomas @teukolsky @tomwlodarczyk @trami18 @wthrowe:

This PR adds you to the author list of the SpECTRE DOI. Please make sure your information in the `Metadata.yaml` file is correct. If it's correct you don't need to comment on this PR. If any of this information changes you can open a PR to edit it at any time and the changes will be reflected in the next release.

wthrowe commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+
+      - 4-5 major infrastructure contributions, each taking roughly 4-5 weeks of
+        full-time work for an existing Core developer. Or, ∼20-25 weeks of
+        full-time infrastructure work saved for a current Core developer.
```
I don't understand this sentence.

nilsvu replied:
@nilsdeppe you originally wrote this in the RFC document. Could you clarify what you meant?

nilsvu replied:
Removed this sentence for now. If you want me to add a clarified version back in, just let me know.

wthrowe commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+      For the time a Developer serves as Core developer they take on some
+      project-management responsibilities, such as merging pull-requests.
+
+      Guidelines on becoming a Core developer:
```
This list reads very awkwardly to me.  I suggest this header be changed to "Guidelines on Core developers:"...

nilsvu replied:
I reworded the list slightly, though didn't change the header because I think it would be good to keep it consistent with the other headers. Happy to adjust further if you think it's necessary.

wthrowe replied:
I don't like the header because the fourth bullet is not a guideline on becoming a core developer.

This whole list still reads very awkwardly.  (The other sections also somewhat, but less so than here.)  I think it is the mixture of imperative and declarative bullets.

nilsvu replied:
Ok, restructured the description a bit. How do you like it now?

wthrowe replied:
This reads much better, thanks.  I think we should keep the "Most people will not end up in the Core developer..." part somewhere if we're still planning for developer tier to have authorship implications, though.

nilsvu replied:
I thought the paragraph above makes the authorship implications sufficiently clear:

> We maintain three tiers of authors for the code as a whole: Core, Developers and Contributors. The full author list for the SpECTRE DOI is built by concatenating all authors in the order they are listed here.
>
> The primary goal of this author list is to recognize infrastructure work that does not or cannot have any science papers and results associated with it. Through authorship on the SpECTRE DOI such work gains recognition that holds value in academia. Authorship on the SpECTRE DOI does not directly entitle a person for authorship rights on other publications. Authorship rights on publications related to SpECTRE are governed by our publication policy.

Once we add our publication policy to the repo we can add a link here. Writing the publication policy is a bit out of scope for this PR, and we should do that after the SXS policies are finished.

Don't you think this is enough explanation for now?

wthrowe replied:
I think that's enough about the policy, but as I said, if developer tier has authorship implications (which I now see that it still does), then I think we should retain the part about most people not becoming core developers.  It's different from the other tiers, where moving up to the tier is expected to happen eventually for anyone continuously working on SpECTRE.

nilsdeppe replied:
I agree with @wthrowe, especially because as time goes on fewer and fewer people will move to Core Devs simply because there will be a lot less infrastructure work to do.

nilsvu replied:
@nilsdeppe @wthrowe added back the paragraph, please check the fixup

wthrowe commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+
+      Guidelines on becoming a Core developer:
+
+      - 4-5 major infrastructure contributions, each taking roughly 4-5 weeks of
```
...this bullet be prefixed with "A core developer has"...

wthrowe commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+      - 4-5 major infrastructure contributions, each taking roughly 4-5 weeks of
+        full-time work for an existing Core developer. Or, ∼20-25 weeks of
+        full-time infrastructure work saved for a current Core developer.
+      - Seeks out PRs to review and volunteers when calls for a reviewer are
```
...and this bullet be prefixed with "A core developer".

wthrowe commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+      Guidelines on becoming a Developer:
+
+      - Do some code review, does not need to be self-initiated.
+      – Somewhere around 40 commits and 10k-15k lines of code should trigger a
```
Inconsistent bullet dash-like thing.

nilsvu replied:
Ahh thanks for spotting this, I had weird formatting issues and didn't understand what caused that..

wthrowe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def simplify_name(name):
+    """Converts the `name` to lower-case ASCI for fuzzy comparisons."""
```
"ASCII"

wthrowe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,179 @@
+import yaml
+
+VERSION_PATTERN = r'(\d{4})\.(\d{2})\.(\d{2})(\.\d+)?'
+DOI_PATTERN = r'10.\d{4,9}/zenodo.(\d+)'
```
Should the two `.` be escaped here?

nilsvu replied:
No, it only needs escaping in a character set like `[\.]`. (I'll double-check this)

nilsvu replied:
Ohh it's the other way 'round 🤦‍♂️ `[.]` needs no escaping.

wthrowe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,179 @@
+            split_name = name.split(', ')
+            self.assertTrue(
+                len(split_name) == 2,
+                f"Name '{name}' should be formatted 'Last name, First name'")
```
Should we handle "Jr.", etc.?  (BibTeX uses "Last, Jr., First".)

nilsdeppe replied:
I think it would be great if we could offer it, but Zenodo only supports `Last Name, First Name` (or at least that's what their documentation says...). However, given that the Zenodo documentation isn't always up-to-date I think it would be good to test to see what happens if we just add a contributor "Last, Jr., First" and see what Zenodo does in the sandbox. @nilsleiffischer would you mind testing that?

nilsvu replied:
Seems fine, at least nothing obviously breaks. In fact it seems like Zenodo doesn't process the name at all, it just appears in the `Last[, Jr.], First` format in the author list of the publication and in the citation suggestions, including the BibTeX entry. I adjusted the test to allow for the optional name component.

nilsdeppe replied:
Okay, awesome! Thanks so much :)

wthrowe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,179 @@
+            check_name(author_name)
+            self.assertEqual(
+                all_author_names_simplified.count(simplify_name(author_name)),
+                1, f"Duplicate author: {author_name}")
```
The "duplicate affiliation" check above is using a different syntax.  I don't know if it matters, but might as well make these all consistent.

wthrowe commented on docs/Doxyfile.in:
```diff
@@ -23,7 +23,11 @@ MULTILINE_CPP_IS_BRIEF = YES
 TAB_SIZE               = 2
 
-ALIASES                = "semantics=<p></p><i>Semantics:</i>" \
+ALIASES                = "spectrename=@SPECTRE_NAME@" \
```
Do we really want a doc substitution for the project name?  Everything, including this PR, hardcodes it, and I don't see a reason to encourage people to do otherwise.

nilsvu replied:
Agreed and removed

macedo22 commented on Metadata.yaml:
```diff
@@ -0,0 +1,246 @@
+        Orcid: 0000-0002-1962-680X
+        Affiliations: [*caltech]
+        GitHub: dongjun826
+      - Name: Macedo, Alex
```
This is quite a pedantic comment, but if possible, could "Alex" be changed to "Alexandra" to reflect my full first name? Wasn't sure if an edit to the contributors' information belonged in this PR or in our own PR after this release. If it's the latter, feel free to ignore, and I can submit a PR for this later!

nilsvu replied:
That's not pedantic at all, of course I'll change that!

trami18 replied:
Piggy backing on what Alex said, would it be okay if we change "Teresa" to my full name "Teresita" as well? :)

nilsvu replied:
Of course, should be fixed now!

trami18 replied:
Thank you Nils! :)

nilsdeppe replied:
I'll echo what @nilsleiffischer: asking us to spell your name the way you want it spelled is not at all pedantic :)

macedo22 replied:
Thank you!

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,245 @@
+  Core:
+    Description: |
+      For the time a Developer serves as Core developer they take on some
+      project-management responsibilities, such as merging pull-requests.
```
I'd prefer something that takes more work than merging PRs (clicking a button on a website isn't exactly a lot of work....) as the example. How about `initiating code reviews, finding and pinging reviewers, maintaining compatibility with third party libraries, and maintaining the CI infrastructure.`

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,245 @@
+        for at least as long as they were already in the Core tier before they
+        left, or for three years, whichever is longer. At that point the Core
+        developers will discuss if the person should be moved to the Developer
+        tier, where they will stay indefinitely. A Core developer can of course
```
I'd remove "of course also" since that's more of a colloquialism.

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,245 @@
+  Developers:
+    Description: |
+      Developers have contributed major features to SpECTRE and remain in this
+      list indefinitely (unless they wish to be removed).
```
I'd remove the parentheses around `unless they wish to be removed`

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,245 @@
+  Contributors:
+    Description: |
+      Anyone who has contributed to SpECTRE will be added to this list and
+      remain a Contributor indefinitely (unless they wish to be removed).
```
I'd remove the parentheses here too

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,245 @@
+        a single PR, all persons involved will become a Contributor.
+
+      This author list is ordered by last name.
+
```
[nitpick] in the above Developer and Core parts you do not have a blank line here

nilsdeppe commented on README.md:
```diff
@@ -21,6 +23,23 @@ eXtreme Spacetimes (SXS) research program into the multi-messenger
+
+- [Find BibTeX entry for this version on Zenodo](https://zenodo.org/record/4290405/export/hx)
+
+To aid reproducibility of your scientific results with SpECTRE we recommend you
```
I think there should be a comma after `SpECTRE`?

nilsdeppe commented on README.md:
```diff
@@ -21,6 +23,23 @@ eXtreme Spacetimes (SXS) research program into the multi-messenger
+
+To aid reproducibility of your scientific results with SpECTRE we recommend you
+keep track of the version(s) you used and report this information in your
+publication. Also supply the YAML input files and, if appropriate, any
```
Add "We also recommend you supply... as supplemental material to the publication."

nilsvu replied:
Is "supplemental" and "supplementary" equally correct?

nilsdeppe commented on cmake/SpectreLoadMetadata.cmake:
```diff
@@ -0,0 +1,57 @@
+# - SPECTRE_ZENODO_ID: ZenodoId
+
+file(READ "${CMAKE_SOURCE_DIR}/Metadata.yaml" SPECTRE_METADATA_FILE)
+# Split lines into cmake list (see https://cmake.org/pipermail/cmake/2007-May/014222.html)
```
[optional] I don't think you need to cite the mailing list. We do this sort of thing in several other locations in the code already

nilsdeppe commented on cmake/SpectreLoadMetadata.cmake:
```diff
@@ -0,0 +1,57 @@
+
+# Read data from the file
+set(SPECTRE_NAME "")
+set(SPECTRE_VERSION "")
```
Should maybe also parse out the homepage for completeness? We might want to print that at startup of executables too. I'm not sure if we use the homepage anywhere in the docs, but conceivably that would also be easier to maintain if we have it in CMake

nilsdeppe replied:
And maybe also add a `GitRepo` that we parse out? I can see that being used in a few places (like in this PR :D )

nilsvu replied:
Added the `SPECTRE_HOMEPAGE` to cmake.

The metadata already has the `GitHub: sxs-collaboration/spectre` field. When parsing it out I couldn't get Doxygen to build a URL like `https://github.com/\spectregithub/xyz` though. The other place where we use the repo name is in the CI workflow files, which we don't CMake-configure, so can't insert the variable. So I'd just leave as-is. If other use-cases turn up we can always parse more metadata into cmake when needed.

nilsdeppe replied:
sounds good

nilsdeppe commented on docs/Installation/VersioningAndReleases.md:
```diff
@@ -0,0 +1,85 @@
+We provide **no guarantee** that releases satisfy any notion of "compatibility"
+with each other. In particular, the interface of any C++ library, Python module
+or executable may change between releases. When writing code for SpECTRE we
+recommend to regularly rebase on the latest `develop` branch of the
```
`to regularly rebase` -> `regularly rebasing`?

nilsdeppe commented on docs/Doxyfile.in:
```diff
@@ -23,7 +23,10 @@ MULTILINE_CPP_IS_BRIEF = YES
 
-ALIASES                = "semantics=<p></p><i>Semantics:</i>" \
+ALIASES                = "spectreversion=@SPECTRE_VERSION@" \
+                         "spectredoi=@SPECTRE_DOI@" \
```
could add `spectrehompage` and `spectregitrepo` here too

nilsvu replied:
See https://github.com/sxs-collaboration/spectre/pull/2606#discussion_r533227169

nilsdeppe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,184 @@
+    def setUpClass(cls):
+        # Always work with the repository that contains this file
+        cls.repo = git.Repo(__file__, search_parent_directories=True)
+        with open(os.path.join(cls.repo.working_dir, 'Metadata.yaml'),
```
should there be some nice error message if this fails? E.g. because we are running from Formaline (or a github version), or we are not running from within the source tree at all?

nilsvu replied:
It raises a Py exception, so it already prints a relatively nice error message:

```
Traceback (most recent call last):
  File "tools/CheckMetadata.py", line 31, in setUpClass
    cls.repo = git.Repo(__file__, search_parent_directories=False)
  File "/Users/nlf/.virtualenvs/spectre/lib/python3.8/site-packages/git/repo/base.py", line 181, in __init__
    raise InvalidGitRepositoryError(epath)
git.exc.InvalidGitRepositoryError: tools/CheckMetadata.py
```

Added a bit more info to the error message now though.

nilsdeppe replied:
okay. Is there ever any need to run this on a release or formaline output? If so we could recursively search up directories for a `Metadata.yaml` if we can't find a git repo

nilsvu replied:
I can't think of a reason this needs to run outside the repo

nilsvu commented:
@wthrowe @nilsdeppe pushed fixups, please take a look

nilsdeppe commented on Metadata.yaml:
```diff
@@ -82,23 +82,31 @@ Authors:
       For the time a Developer serves as Core developer they take on some
-      project-management responsibilities, such as merging pull-requests.
+      project-management responsibilities, such as initiating code reviews,
+      merging pull-requests, maintaining compatibility with third-party
```
I would like you to remove "merging pull requests"

nilsvu replied:
reworded the fixup

nilsdeppe replied:
Looks good, thanks!

nilsdeppe commented:
I'm happy for you to squash once @wthrowe has had a chance to look :)

wthrowe commented on docs/Installation/VersioningAndReleases.md:
```diff
@@ -29,7 +29,7 @@ Each release is **guaranteed** to pass all automated tests laid out in the
-recommend to regularly rebase on the latest `develop` branch of the
+recommend regularly rebasing on the latest `develop` branch of the
 [`sxs-collaboration/spectre`](https://github.com/sxs-collaboration/spectre)
 repository and aim to contribute your code upstream by following our \ref
```
"aim" -> "aiming" to match change to "rebasing"

nilsvu commented:
@nilsdeppe @wthrowe could you take another look at the last fixup commit and let me know when to squash? (Even if we want to make further changes I'd suggest I squash the fixups up to this point)

nilsdeppe commented on cmake/SpectreLoadMetadata.cmake:
```diff
@@ -0,0 +1,66 @@
+
+# Read the following information from the `Metadata.yaml` file:
+# - SPECTRE_NAME: Name
+# - SPECTRE_VERSION: Version
```
need to add `SPECTRE_HOMEPAGE` here

nilsdeppe reviewed (dismissed):
Looks good. I'm happy for you to squash, including this one minor thing. I'll do a final comparison to what we have in the other repo tomorrow.

nilsvu commented:
@haraldp271 Based on the added guidelines for supervisors I added you to the author list. Please make sure your information is correct (no need to comment if everything is correct).

wthrowe commented:
Looks good.  Squash.

nilsvu commented:
Thanks @wthrowe. Squashed up the the point where I added the supervision guidelines. @nilsdeppe and @kidder Could you take another look? I'll take this PR out of draft mode tomorrow morning (CET 😉). Then I think everyone had enough time to look at their data and we can merge whenever we're ready :)

nilsdeppe commented on cmake/SpectreLoadMetadata.cmake:
```diff
@@ -0,0 +1,67 @@
+# Check we have read all required keys
+if(NOT SPECTRE_NAME)
+  message(
+    FATAL_ERROR "Could not determine project name. Please check the "
```
I wonder if these errors should be expanded to something like `Please check that the Metadata.yaml file exists and has a line that would match the regular expression ^Name: (.*)? You can double check the regular expression in the SpectreLoadMetadata.cmake file."? I assume we will hit this error rarely, if ever, but that probably means a more descriptive error message is better.

nilsdeppe commented on src/PythonBindings/setup.py:
```diff
@@ -10,7 +10,7 @@
     description="Python bindings for SpECTRE",
     author="SXS collaboration",
-    url="https://spectre-code.org",
+    url="${SPECTRE_HOMEPAGE}",
```
I don't understand how the version and url work. Where does this ${....} come from? Or is this inserting something into CMake? Some documentation in the sense of code comments would be helpful.

nilsvu replied:
Either `@var@` or `${var}` work with cmake https://cmake.org/cmake/help/latest/command/configure_file.html. Changed to `@`-notation since we use that more.

nilsdeppe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,189 @@
+        self.assertEqual(
+            sorted(authors_devs_list,
+                   key=lambda a: a['Name']), authors_devs_list,
+            "'Developers' author list should be ordered alphabetically")
```
Is there a good way to list which author(s) are not in order?

Also, `alphabetically by last name` (just to be extra clear)

Same below

nilsvu replied:
The `self.assertEqual` prints that information by default:

```
======================================================================
FAIL: test_authors (__main__.TestMetadata)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tools/CheckMetadata.py", line 136, in test_authors
    self.assertEqual(
AssertionError: Lists differ: [{'Na[1762 chars]e': 'Scheel, Mark A.', 'Orcid': '0000-0001-665[531 chars]ky'}] != [{'Na[1762 chars]e': 'Teukolsky, Saul A.', 'Orcid': '0000-0001-[531 chars]el'}]

First differing element 8:
{'Name': 'Scheel, Mark A.', 'Orcid': '0000-0001-665[182 chars]eel'}
{'Name': 'Teukolsky, Saul A.', 'Orcid': '0000-0001-[290 chars]sky'}

Diff is 3340 characters long. Set self.maxDiff to None to see it. : 'Developers' author list should be ordered alphabetically by last name
```

nilsdeppe commented on tools/CheckMetadata.py:
```diff
@@ -0,0 +1,189 @@
+
+VERSION_PATTERN = r'(\d{4})\.(\d{2})\.(\d{2})(\.\d+)?'
+DOI_PATTERN = r'10\.\d{4,9}/zenodo\.(\d+)'
+ORCID_PATTERN = r'\d{4}-\d{4}-\d{4}-\d{3}[0-9X]'
```
I haven't look, but are we guaranteed ORCID won't add more letters? If not, maybe in the check below add to the error message that it could be that more letters should be added?

nilsvu replied:
Yea the letter is just the checksum:

https://support.orcid.org/hc/en-us/articles/360006897674-Structure-of-the-ORCID-Identifier

> The last character in the ORCID iD is a checksum. In accordance with ISO/IEC 7064:2003, MOD 11-2, this checksum must be "0" - "9" or "X", a capital letter X which represents the value 10.

nilsdeppe commented on docs/Installation/VersioningAndReleases.md:
```diff
@@ -0,0 +1,85 @@
+  processed or converted to newer formats.
+- Python bindings: We try to offer deprecation warnings when changing the Python
+  interfaces so developers of external packages that use the Python bindings
+  can keep up.
```
maybe replace `keep up` with `have time to update their packages.`?

nilsdeppe commented on Metadata.yaml:
```diff
@@ -0,0 +1,278 @@
+        etc. The supervision is what will be counted for contributor status.
+        Generally, a paper is written where the person who came up with the idea
+        is a co-author, which is how credit is given for the idea. An example
+        would be suggesting "We should try reconstruction scheme X" without
```
maybe `would be` -> `is`? I found the occurrence of `would` twice in this sentence makes it a bit more difficult to read than necessary. (My fault, I think I wrote this)

nilsdeppe reviewed (commented):
last few minor suggestions. I've gone through everything in detail and verified ORCIDs match the PDF. Please go ahead and squash immediately :)

nilsvu commented:
@moble I added you to the Contributors author list. Please make sure your information is correct (no need to comment if everything is correct).

nilsvu commented:
@10220 @carmaza @denyzamelchor @dongjun826 @ffoucart @fmahebert @geoffrey4444 @gsb76 @haraldp271 @hrueter @isaaclegred @isha1810 @jennifersanchez @kidder @macedo22 @markscheel @MarloMo @mgiesler @moble @moxcodes @nilsdeppe @osheamonn @prayush @sierracthomas @teukolsky @tomwlodarczyk @trami18 @wthrowe:

Your authorship information is now merged into the repository and published to Zenodo at this DOI: https://zenodo.org/record/4290405. You'll be able to edit your information by opening a pull-request at any time and it will propagate to Zenodo [at the next release, approximately monthly](https://spectre-code.org/versioning_and_releases.html). You can also [cite the SpECTRE DOI](https://spectre-code.org/) in papers now!
